stdlib/seq.mli

```diff
@@ -13,97 +13,760 @@
+    the element [f x y] appears once as an element of [map_product f xs ys].
+
+    The order in which these elements appear is unspecified.
+
+    The sequences [xs] and [ys] are not required to be finite.
+
+    The sequences [xs] and [ys] must be persistent.
+
+    [map_product f xs ys] is equivalent to
+    [map (fun (x, y) -> f x y) (product xs ys)].
+
+    @since 4.14 *)
+
+(** {1 Splitting a sequence into two sequences} *)
+
+val unzip : ('a * 'b) t -> 'a t * 'b t
+(** [unzip] transforms a sequence of pairs into a pair of sequences.
+
+    [unzip xs] is equivalent to [(map fst xs, map snd xs)].
+
+    Querying either of the sequences returned by [unzip xs]
+    causes [xs] to be queried.
+    Therefore, querying both of them
+    causes [xs] to be queried twice.
+    Thus, [xs] must be persistent and cheap.
+    If that is not the case, use [unzip (memoize xs)].
+
+    @since 4.14 *)
+
+val split : ('a * 'b) t -> 'a t * 'b t
+(** [split] is an alias for [unzip].
+
+    @since 4.14 *)
+
+val partition_map : ('a -> ('b, 'c) Either.t) -> 'a t -> 'b t * 'c t
+(** [partition_map f xs] returns a pair of sequences [(ys, zs)], where:
+
+    - [ys] is the sequence of the elements [y] such that
+      [f x = Left y], where [x] ranges over [xs];
+
+    - [zs] is the sequence of the elements [z] such that
+      [f x = Right z], where [x] ranges over [xs].
+
+    [partition_map f xs] is equivalent to a pair of
+    [filter_map Either.find_left (map f xs)] and
+    [filter_map Either.find_right (map f xs)].
+
+    Querying either of the sequences returned by [partition_map f xs]
+    causes [xs] to be queried.
+    Therefore, querying both of them
+    causes [xs] to be queried twice.
+    Thus, [xs] must be persistent and cheap.
+    If that is not the case, use [partition_map f (memoize xs)].
+
+    @since 4.14 *)
+
+val partition : ('a -> bool) -> 'a t -> 'a t * 'a t
+(** [partition p xs] returns a pair of the subsequence of the elements
+    of [xs] that satisfy [p] and the subsequence of the elements of
+    [xs] that do not satisfy [p].
+
+    [partition p xs] is equivalent to
+    [filter p xs, filter (fun x -> not (p x)) xs].
+
+    Consuming both of the sequences returned by [partition p xs] causes
+    [xs] to be consumed twice and causes the function [f] to be applied
+    twice to each element of the list.
+    Therefore, [f] should be pure and cheap.
+    Furthermore, [xs] should be persistent and cheap.
+    If that is not the case, use [partition p (memoize xs)].
+
+    @since 4.14 *)
+
+(** {1 Converting between sequences and iterators} *)
+
+(** An iterator is a representation of a sequence as a function of type
+    [unit -> 'a option]. Every time this function is invoked, it returns
+    the next element of the sequence. When there are no more elements,
+    it returns [None]. An iterator has mutable internal state, therefore
+    is ephemeral: the sequence that it represents can be consumed at most
+    once. *)
```

Comment:
I would also be in favor of not using these names.

Reply:
I understand that "iterator" is indeed very much an overloaded word. I suggest using "distributor" instead, which should be more neutral (as far as I know, this word has no fixed meaning in a programming context). I have committed the change.

Regarding exceptions versus options, I don't think efficiency should be the deciding factor (it is compiler-dependent anyway). I tend to favor options because they are safer; the user cannot overlook the fact that they must handle the case where there are no more elements.

Reply:
Thanks @CraigFe for your review (and quick reaction)!